vertical-pod-autoscaler/e2e/updater.go

```diff
@@ -152,48 +320,91 @@ func setupPDB(f *framework.Framework, name string, maxUnavailable int) *policyv1
 	return pdb
 }
 
-func makePodSet(pods *apiv1.PodList) map[string]bool {
-	result := make(map[string]bool)
+type podSet map[string]types.UID
+
+func makePodSet(pods *apiv1.PodList) podSet {
+	result := make(podSet)
+	if pods == nil {
+		return result
+	}
 	for _, p := range pods.Items {
-		result[p.Name] = true
+		result[p.Name] = p.UID
 	}
 	return result
 }
 
-func waitForPodSetChangedInDeployment(c clientset.Interface, deployment *appsv1.Deployment, podList *apiv1.PodList) error {
+func getCurrentPodSetForDeployment(c clientset.Interface, d *appsv1.Deployment) podSet {
+	podList, err := framework.GetPodsForDeployment(c, d)
+	gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	return makePodSet(podList)
+}
+
+func waitForPodSetChanged(c clientset.Interface, getCurrentPodList getCurrentPodListFunc,
+	podList *apiv1.PodList) error {
 	initialPodSet := makePodSet(podList)
 
 	err := wait.PollImmediate(pollInterval, pollTimeout, func() (bool, error) {
-		return isPodSetChanged(c, deployment, initialPodSet)
-
+		currentPodList, err := getCurrentPodList()
+		if err != nil {
+			return false, err
+		}
+		currentPodSet := makePodSet(currentPodList)
+		return isPodSetChanged(c, currentPodSet, initialPodSet), nil
 	})
 
 	if err != nil {
-		return fmt.Errorf("Waiting for set of pods changed in %v: %v", deployment.Name, err)
+		return fmt.Errorf("Waiting for set of pods changed: %v", err)
 	}
 	return nil
 }
 
-func isPodSetChanged(c clientset.Interface, deployment *appsv1.Deployment, initialPodSet map[string]bool) (bool, error) {
-	count, err := getPodSetChanges(c, deployment, initialPodSet)
-	if err != nil {
-		return false, err
-	}
-	return count > 0, nil
+func isPodSetChanged(c clientset.Interface, currentPodSet podSet, initialPodSet podSet) bool {
+	return getPodSetChanges(c, currentPodSet, initialPodSet) > 0
 }
 
-func getPodSetChanges(c clientset.Interface, deployment *appsv1.Deployment, initialPodSet map[string]bool) (int, error) {
-	currentPodList, err := framework.GetPodsForDeployment(c, deployment)
-	if err != nil {
-		return 0, err
-	}
-	currentPodSet := makePodSet(currentPodList)
+func getPodSetChanges(c clientset.Interface, currentPodSet podSet, initialPodSet podSet) int {
 	diffs := 0
-	for name, inInitial := range initialPodSet {
-		inCurrent := currentPodSet[name]
-		if inInitial && !inCurrent {
+	for name, initialUID := range initialPodSet {
+		currentUID, inCurrent := currentPodSet[name]
+		if !inCurrent {
 			diffs += 1
+		} else if initialUID != currentUID {
+			diffs += 1
+		}
+	}
+	return diffs
+}
+
+func createReplicaSetWithRetries(c clientset.Interface, namespace string, obj *appsv1.ReplicaSet) error {
+	if obj == nil {
+		return fmt.Errorf("Object provided to create is empty")
+	}
+	createFunc := func() (bool, error) {
+		_, err := c.AppsV1().ReplicaSets(namespace).Create(obj)
+		if err == nil || apierrs.IsAlreadyExists(err) {
```

Comment:
I can see that this is a general way to create test resources but aren't we afraid that an existing ReplicaSet could signify an error (conflicting definitions could or could not influence our test)?

Reply:
There should be no conflicting definitions. Even if cleanup from a previous test fails, every test sets up a new namespace.